feat(staking): LW-7674 split delegation&(de)registration tx into two activities

xdzurman opened the pull request:
### **Recommended to review by commits**

# Checklist

- [x] [JIRA](https://input-output.atlassian.net/jira/software/c/projects/LW/boards/1209?modal=detail&selectedIssue=LW-7674&assignee=5cfdfc4bc97d8b0c633413a7&assignee=UNASSIGNED_USER_ID)
- [ ] Proper tests implemented
- [x] Screenshots added.

---

## Proposed solution

Split the transaction, which both delegates and de-registers keys, into two actions. 
- Delegation action shows tx fee, and no returned deposits
- De-registration action shows only returned deposits (not fees), so that fees are not duplicated and deposits are nice round number

This applies to pending txs as well. 
One edge case is when user re-delegates from 2 pools (A, B) into one pool (A) - this is not a delegation action chain-wise, only a de-registration. In such a case, we cannot split it into 2 actions, since the delegation action isn't anywhere on-chain. Therefore, this action will be shown as just "De-registration" and its value is deposit deducted by tx fee (e.g. 1.77 ADA = 2 ADA - 0.23 ADA) 

Since there is no plus or minus of values in the activity tab, I introduced a `Returned deposit` label in `Transaction Detail` to make it clear for the user. Tech-wise, this avoids a lot of mental gymnastics we'd have to deal with if both paid and returned deposits were handled by just one prop (now we have both `deposit` and `returnedDeposit` props)

## Testing

- Stake from two pools (A, B) into one pool that is different from them (C) - you should see 1 Deregistration and 1 Delegation activity
- Stake from two pools (A, B) into one pool that is one from them (A) - you should see 1 Deregistration activity
- Stake to more pools than previously - you should see 1 Registration and 1 Delegation activity

## Screenshots

![image](https://github.com/input-output-hk/lace/assets/22454337/f58dff77-ba7e-433c-b41a-f0ba7c973db4)
![image](https://github.com/input-output-hk/lace/assets/22454337/afb1a4c4-6aef-4fee-b354-f2fc1e87c639)

Detail of delegation activity:
![image](https://github.com/input-output-hk/lace/assets/22454337/d2fc40a3-d907-4db2-b924-01a7e1cbeb1b)
Detail of de-registration activity:
![image](https://github.com/input-output-hk/lace/assets/22454337/41abb68b-1055-4692-a3f7-0f6039f1bf3b)

Registration & Delegation tx:
![image](https://github.com/input-output-hk/lace/assets/22454337/61b412da-f379-45c5-b849-5a2e524c2082)






<!-- allure -->
---
# Allure report
`allure-report-publisher` generated test report!

<!-- jobs -->
<!-- smokeTests -->
**smokeTests**: ❌ [test report](https://lace-qa:8PP5wBaDV6UoXj@dq4ajm5i5q7bz.cloudfront.net/linux/chrome/1681/5964412553/index.html) for [0f83df6b](https://github.com/input-output-hk/lace/pull/433/commits/0f83df6b523eb7e712064490b1c7a72e0dde3d0b)
|       | passed | failed | skipped | flaky | total | result |
|-------|--------|--------|---------|-------|-------|--------|
| Total | 34     | 1      | 0       | 0     | 35    | ❌     |
<!-- smokeTests -->

<!-- jobs -->
<!-- allurestop -->

xdzurman commented on packages/core/src/ui/components/Activity/AssetActivityItem.tsx:
```diff
@@ -95,7 +96,7 @@ const TransactionStatusIcon = ({ status, type }: TransactionStatusIconProps) =>
 };
 
-const transaltionTypes = {
+const translationTypes = {
```
typo

xdzurman commented on packages/core/src/ui/components/Transactions/Transaction.tsx:
```diff
@@ -311,17 +339,6 @@ export const Transaction = ({
           </div>
         </div>
       )}
-      {deposit && (
```
Just moved this above "Inputs" and "Outputs" in tx detail according to figma designs.

xdzurman commented on packages/core/src/ui/components/Transactions/Transaction.tsx:
```diff
@@ -269,6 +274,29 @@ export const Transaction = ({
+            </div>
+          </div>
+        )}
+        {returnedDeposit && (
```
Currently these are "two separate rows", so perhaps some logic for exclusion could be added, but there should never be a situation where both `deposit` and `returnedDeposit` are in one transaction

xdzurman replied:
removed altogether, it wasn't necessary

przemyslaw-wlodek commented on apps/browser-extension-wallet/src/stores/slices/wallet-activities-slice.ts:
```diff
@@ -53,7 +53,14 @@ const DelagatioinTransactionTypes = new Set(['delegation', 'delegationRegistrati
 
 const getDelegationAmount = (activity: AssetActivityItemProps) => {
```
Could you please fix the type here to include only delegation types? (`DelagatioinTransactionTypes` - with typo in codebase)

przemyslaw-wlodek commented on apps/browser-extension-wallet/src/stores/slices/wallet-activities-slice.ts:
```diff
@@ -53,7 +53,14 @@ const DelagatioinTransactionTypes = new Set(['delegation', 'delegationRegistrati
 const getDelegationAmount = (activity: AssetActivityItemProps) => {
   const fee = new BigNumber(Number.parseFloat(activity.fee));
-  return activity?.deposit ? fee.plus(activity.deposit) : fee;
+  return (
+    fee
+      .plus(activity.deposit || 0)
+      .minus(activity.returnedDeposit || 0)
+      // Currently, all values in activity screen are absolute
+      // Remove .abs() when this is changed
+      .abs()
+  );
```
Can we be more explicit in this case? It requires a deep understanding of the flow to know that both `deposit` and `returnedDeposit` cannot be present at the same time.

After applying the above comment we can do the following:

```suggestion
  if (activity.type === 'delegationRegistration') {
    return fee.plus(activity.deposit);
  }
  
  if (activity.type === 'delegationDeregistration') {
    return new BigNumber(activity.returnedDeposit).minus(fee);
  }
  
  return fee;
```

przemyslaw-wlodek replied:
I hope that when you narrow the types you will mitigate unhandled edge cases + the QA phase will reveal if something is there. Technically it'd be more correct to do it this way, as we described but given the time pressure and that your logic also "does the job" 😄 it's for you to decide.

przemyslaw-wlodek reviewed (approved):
Great work! We've left some improvement proposals.

przemyslaw-wlodek reviewed (changes requested):
Sorry, I wanted to request changes in the previous review 😅

xdzurman commented on apps/browser-extension-wallet/src/stores/slices/wallet-activities-slice.ts:
```diff
@@ -101,56 +108,36 @@ const getWalletActivitiesObservable = async ({
+        setTransactionDetail(
+          tx,
+          transformedTx.direction,
+          transformedTx.status || Wallet.TransactionStatus.SUCCESS,
```
Before there was hardcoded `Wallet.TransactionStatus.SUCCESS` so I wasn't sure whether this can fail somehow, I can remove it however and try to trust it

xdzurman commented:
I split the delegation & Registration tx into two actions in this PR, since the changes are almost identical. Note that I also changed the order so that Registration/Deregistration appears "older in time", and so that the Delegation comes on top as the "main" action of the two.
![image](https://github.com/input-output-hk/lace/assets/22454337/9b0ca6c0-c114-4099-8163-0ac549dc477d)

szymonmaslowski commented on apps/browser-extension-wallet/src/stores/slices/wallet-activities-slice.ts:
```diff
@@ -105,56 +126,31 @@ const getWalletActivitiesObservable = async ({
       cardanoCoin
     });
 
+    const finalizeTransaction = (transformedTx: Omit<AssetActivityItemProps, 'onClick'>) => ({
```
```suggestion
    const extendWithClickHandler = (transformedTx: Omit<AssetActivityItemProps, 'onClick'>) => ({
```

or at least `finalizeTransactionEntryData`

`finalizeTransaction` sounds to me like it would be submiting the transaction 😉

szymonmaslowski commented on packages/core/src/ui/components/Transactions/Transaction.tsx:
```diff
@@ -269,6 +274,29 @@ export const Transaction = ({
           </div>
         </div>
+
+        {deposit && (
+          <div className={styles.details}>
+            <div className={styles.title}>{t('package.core.transactionDetailBrowser.deposit')}</div>
+            <div className={styles.detail}>
+              <div className={styles.amount}>
+                <span className={styles.ada}>{`${deposit} ${coinSymbol}`}</span>
+                <span className={styles.fiat}>{amountTransformer(deposit)}</span>
+              </div>
+            </div>
+          </div>
+        )}
+        {depositReclaim && (
+          <div className={styles.details}>
+            <div className={styles.title}>{t('package.core.transactionDetailBrowser.depositReclaim')}</div>
+            <div className={styles.detail}>
+              <div className={styles.amount}>
+                <span className={styles.ada}>{`${depositReclaim} ${coinSymbol}`}</span>
+                <span className={styles.fiat}>{amountTransformer(depositReclaim)}</span>
+              </div>
+            </div>
+          </div>
+        )}
```
I would reduce code duplication here
```suggestion
        const renderDepositValueSection = ({ value, label }) => (
          <div className={styles.details}>
            <div className={styles.title}>{label}</div>
            <div className={styles.detail}>
              <div className={styles.amount}>
                <span className={styles.ada}>{`${value} ${coinSymbol}`}</span>
                <span className={styles.fiat}>{amountTransformer(value)}</span>
              </div>
            </div>
          </div>
        )
        ...
        {deposit && renderDepositValueSection({ value: deposit, label: t('package.core.transactionDetailBrowser.deposit') })}
        {depositReclaim && renderDepositValueSection({ value: depositReclaim, label: t('package.core.transactionDetailBrowser.depositReclaim') })}
```

szymonmaslowski reviewed (approved):
I left few minor suggesctions 😉
